### PR TITLE
Fix tracing send on closed channel

### PR DESCRIPTION
**Pull request by klauspost**
## Description

Depending on when the context cancelation is picked up the handler may return and close the channel before `SubscribeJSON` returns, causing:

```
Feb 05 17:12:00 s3-us-node11 minio[3973657]: panic: send on closed channel
Feb 05 17:12:00 s3-us-node11 minio[3973657]: goroutine 378007076 [running]:
Feb 05 17:12:00 s3-us-node11 minio[3973657]: github.com/minio/minio/internal/pubsub.(*PubSub[...]).SubscribeJSON.func1()
Feb 05 17:12:00 s3-us-node11 minio[3973657]:         github.com/minio/minio/internal/pubsub/pubsub.go:139 +0x12d
Feb 05 17:12:00 s3-us-node11 minio[3973657]: created by github.com/minio/minio/internal/pubsub.(*PubSub[...]).SubscribeJSON in goroutine 378010884
Feb 05 17:12:00 s3-us-node11 minio[3973657]:         github.com/minio/minio/internal/pubsub/pubsub.go:124 +0x352
```

Wait explicitly for the goroutine to exit.

Bonus: Listen for `doneCh` when sending to not risk getting blocked there is channel isn't being emptied.

## How to test this PR?

`mc admin trace` on distributed setup.. Probably needs some load to trigger.

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Fixes a regression #18903

### CODE REVIEW

File: internal/pubsub/pubsub.go [hunk 1]
```diff
@@ -136,7 +142,12 @@ func (ps *PubSub[T, M]) SubscribeJSON(mask M, subCh chan<- []byte, doneCh <-chan
 				if err != nil {
 					break
 				}
-				subCh <- append(GetByteBuffer()[:0], buf.Bytes()...)
+
+				select {
+				case subCh <- append(GetByteBuffer()[:0], buf.Bytes()...):
+				case <-doneCh:
+					break loop
```
Comment (harshavardhana) [inline internal/pubsub/pubsub.go]:
```suggestion
return
```

We can just return here right?

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: internal/pubsub/pubsub.go (hunk 1 above)
Comment (klauspost) [reply internal/pubsub/pubsub.go]: There is some logic after the loop... I was annoyed as well.... Maybe we should defer that logic.

Comment (klauspost) [reply internal/pubsub/pubsub.go]: Rewrote it. PTAL

**Review by harshavardhana — APPROVED** [review]

**Comment by harshavardhana** [comment]
This PR fixes the `go.mod` replace tag https://github.com/minio/minio/pull/18984

**Comment by klauspost** [comment]
Tested with several runs.

**Review by vadmeste — APPROVED** [review]
LGTM

**Pull request merged**